core/modules/file/file.module

```diff
@@ -2929,42 +2896,72 @@ function file_stream_wrappers_alter(&$wrappers) {
  * Implements hook_file_access().
  */
 function file_file_access($op, $file, $account) {
-  // If the file URI is invalid, deny access.
-  if (is_object($file) && isset($file->uri) && !file_valid_uri($file->uri)) {
-    if(isset($file->is_new) && $file->is_new == true && user_access('create files', $account)) {
-      return FILE_ACCESS_ALLOW;
-    }
-    return FILE_ACCESS_DENY;
-  }
 
   if ($op == 'create') {
+    // @todo File::createAccess?
     if (user_access('create files', $account)) {
       return FILE_ACCESS_ALLOW;
     }
   }
+  // If the file is not new ('create' got handled above) and its URI is invalid,
+  // we deny access.
+  elseif (!file_valid_uri($file->uri)) {
+    return FILE_ACCESS_DENY;
```

Comment:
Is this check of file_valid_uri needed here? Function file_valid_uri checks whether the uri has a valid scheme (public, private, etc). 

So, if a managed file uri has the wrong scheme, wouldn't that be caught somewhere else? What does a wrong scheme have to do with whether a user has permissions to access a file?

Reply:
> Is this check of file_valid_uri needed here

Try to figure it out with git blame - that code existed before, I just moved it down a bit. Yes, I think it's necessary.

Reply:
Yes, this is legacy from the contrib module `file_entity` [added here](https://www.drupal.org/project/file_entity/issues/1227706) . It does not harm, but it may be redundant. Let's not worry about this one and focus on the other ones.